packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -1066,6 +1071,87 @@ function reappearLayoutEffectsOnFiber(node: Fiber) {
   }
 }
 
+function commitTransitionProgress(
+  finishedRoot: FiberRoot,
+  offscreenFiber: Fiber,
+) {
+  // This function adds suspense boundaries to the root
+  // or tracing marker's pendingSuspenseBoundaries map.
+  // When a suspense boundary goes from a resolved to a fallback
+  // state we add the boundary to the map, and when it goes from
+  // a fallback to a resolved state, we remove the boundary from
+  // the map.
+
+  // We use stateNode on the Offscreen component as a stable object
+  // that doesnt change from render to render. This way we can
+  // distinguish between different Offscreen instances (vs. the same
+  // Offscreen instance with different fibers)
+  const offscreenInstance: OffscreenInstance = offscreenFiber.stateNode;
+
+  let prevState: SuspenseState | null = null;
+  const previousFiber = offscreenFiber.alternate;
+  if (previousFiber !== null && previousFiber.memoizedState !== null) {
+    prevState = previousFiber.memoizedState;
+  }
+  const nextState: SuspenseState | null = offscreenFiber.memoizedState;
+
+  const wasHidden = prevState !== null;
+  const isHidden = nextState !== null;
+
+  const rootState: RootState = finishedRoot.current.memoizedState;
+  // TODO(luna) move pendingSuspenseBoundaries and transitions from
+  // HostRoot fiber to FiberRoot
+  const rootPendingBoundaries = rootState.pendingSuspenseBoundaries;
+  const rootTransitions = rootState.transitions;
+
+  // If there is a name on the suspense boundary, store that in
+  // the pending boundaries.
+  let name = null;
+  const parent = offscreenFiber.return;
+  if (
+    parent !== null &&
+    parent.tag === SuspenseComponent &&
+    parent.memoizedProps.unstable_name
+  ) {
+    name = parent.memoizedProps.unstable_name;
+  }
+
+  if (rootPendingBoundaries !== null) {
+    if (previousFiber === null) {
+      // Initial mount
+      if (isHidden) {
+        rootPendingBoundaries.set(offscreenInstance, {
+          name,
+        });
+      }
+    } else {
+      if (wasHidden && !isHidden) {
+        // The suspense boundary went from hidden to visible. Remove
+        // the boundary from the pending suspense boundaries set
+        // if it's there
+        if (rootPendingBoundaries.has(offscreenInstance)) {
+          rootPendingBoundaries.delete(offscreenInstance);
+        }
+      } else if (!wasHidden && isHidden) {
+        // The suspense boundaries was just hidden. Add the boundary
+        // to the pending boundary set if it's there
+        rootPendingBoundaries.set(offscreenInstance, {
+          name,
+        });
+      }
+    }
+
+    if (rootPendingBoundaries.size === 0 && rootTransitions !== null) {
```

Comment:
Nit: Move this right after the `.delete` call above since logically that's the only case where the count could have reached 0.